src/JSInterop/Microsoft.JSInterop/src/JSRuntimeBase.cs

```diff
@@ -20,40 +22,99 @@ public abstract class JSRuntimeBase : IJSRuntime
 
+        /// <summary>
+        /// Gets or sets the default timeout for asynchronous JavaScript calls.
+        /// </summary>
+        protected TimeSpan? DefaultAsyncCallTimeout { get; set; }
+
         /// <summary>
         /// Invokes the specified JavaScript function asynchronously.
         /// </summary>
         /// <typeparam name="T">The JSON-serializable return type.</typeparam>
         /// <param name="identifier">An identifier for the function to invoke. For example, the value <code>"someScope.someFunction"</code> will invoke the function <code>window.someScope.someFunction</code>.</param>
         /// <param name="args">JSON-serializable arguments.</param>
+        /// <param name="cancellationToken">A cancellation token to signal the cancellation of the operation.</param>
         /// <returns>An instance of <typeparamref name="T"/> obtained by JSON-deserializing the return value.</returns>
-        public Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        public Task<T> InvokeAsync<T>(string identifier, IEnumerable<object> args, CancellationToken cancellationToken = default)
         {
-            // We might consider also adding a default timeout here in case we don't want to
-            // risk a memory leak in the scenario where the JS-side code is failing to complete
-            // the operation.
-
             var taskId = Interlocked.Increment(ref _nextPendingTaskId);
-            var tcs = new TaskCompletionSource<T>();
+            var tcs = new TaskCompletionSource<T>(TaskContinuationOptions.RunContinuationsAsynchronously);
+            if (cancellationToken != default)
+            {
+                _cancellationRegistrations[taskId] = cancellationToken.Register(() =>
+                {
+                    tcs.TrySetCanceled();
+                    CleanupRegistrations(taskId);
+                });
+            }
             _pendingTasks[taskId] = tcs;
 
             try
             {
-                var argsJson = args?.Length > 0 ?
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    tcs.TrySetCanceled();
+                    CleanupRegistrations(taskId);
+
+                    return tcs.Task;
+                }
+
+                var argsJson = args?.Any() == true ?
                     JsonSerializer.Serialize(args, JsonSerializerOptionsProvider.Options) :
                     null;
                 BeginInvokeJS(taskId, identifier, argsJson);
+
                 return tcs.Task;
             }
             catch
             {
-                _pendingTasks.TryRemove(taskId, out _);
+                CleanupRegistrations(taskId);
                 throw;
             }
         }
 
+        private void CleanupRegistrations(long taskId)
+        {
+            _pendingTasks.TryRemove(taskId, out _);
+            _cancellationRegistrations.TryRemove(taskId, out var registration);
```

Comment:
This should  guard against this returning `false`

Reply:
I agree that would be a bit better than having the comment on L86

Reply:
done